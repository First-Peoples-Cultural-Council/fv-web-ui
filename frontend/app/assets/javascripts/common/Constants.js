/*
import {
  //...
} from 'common/Constants'
*/
export const STATE_LOADING = 0 // component is loading (eg: getting data) or busy (eg: submitting data)
export const STATE_DEFAULT = 1 // initial, loaded state, eg: displaying a form
export const STATE_ERROR = 2 // component is not happy, ie: form validation (not .js errors)
export const STATE_SUCCESS = 3 // component is happy, ie: form submitted
export const STATE_SUCCESS_DELETE = 5 // desctructive action has succeeded
export const STATE_ERROR_BOUNDARY = 4 // component has triggered a .js error or is completely messed up

export const DEFAULT_PAGE = 1
export const DEFAULT_PAGE_SIZE = 100
export const DEFAULT_LANGUAGE = 'english'
export const DEFAULT_SORT_COL = 'dc:title'
export const DEFAULT_SORT_TYPE = 'asc'

export const WORKSPACES = 'Workspaces'
export const SECTIONS = 'sections'

export const CONTENT_FULL_WIDTH = 1

export const UNKNOWN = 0
export const WORD = 1
export const PHRASE = 2
export const SONG = 3
export const STORY = 4
