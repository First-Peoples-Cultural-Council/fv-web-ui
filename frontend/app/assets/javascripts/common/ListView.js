import React, { useState, useEffect } from 'react'
import ProviderHelpers from 'common/ProviderHelpers'
import NavigationHelpers, {
  windowLocationPathnameWithoutPagination,
  getSearchObject,
  getSearchObjectAsUrlQuery,
} from 'common/NavigationHelpers'
import Checkbox from 'views/components/Form/Common/Checkbox'
import ConfirmationDelete from 'views/components/Confirmation'

import KeyboardArrowUp from '@material-ui/icons/KeyboardArrowUp'
import KeyboardArrowDown from '@material-ui/icons/KeyboardArrowDown'
import UnfoldMore from '@material-ui/icons/UnfoldMore'

// batchTitle
// ============================================
export const batchTitle = ({ uidsNotDeleted = [], selected, setSelected, copyDeselect, copySelect }) => {
  // uidsNotDeleted = getUidsThatAreNotDeleted({ computedData, deletedUids })
  // All items selected, show deselect
  if (uidsNotDeleted.length === selected.length && uidsNotDeleted.length !== 0) {
    return (
      <button
        className="_btn _btn--compact"
        type="button"
        onClick={() => {
          setSelected([])
        }}
      >
        {copyDeselect}
      </button>
    )
  }
  // show select
  return (
    <button
      className="_btn _btn--compact"
      type="button"
      onClick={() => {
        setSelected(uidsNotDeleted)
      }}
    >
      {copySelect}
    </button>
  )
}
// batchFooter
// ============================================
export const batchFooter = ({
  colSpan,
  confirmationAction,
  copyBtnConfirm,
  copyBtnDeny,
  copyBtnInitiate,
  copyIsConfirmOrDenyTitle,
  selected,
}) => {
  return {
    colSpan,
    element: (
      <ConfirmationDelete
        confirmationAction={confirmationAction}
        className="Contributor__delete"
        compact
        copyIsConfirmOrDenyTitle={copyIsConfirmOrDenyTitle}
        copyBtnInitiate={copyBtnInitiate}
        copyBtnDeny={copyBtnDeny}
        copyBtnConfirm={copyBtnConfirm}
        disabled={selected.length === 0}
      />
    ),
  }
}
// batchRender
// ============================================
export const batchRender = ({ dataUid, selected, setSelected }) => {
  const uid = dataUid
  return (
    <Checkbox
      selected={isSelected({ selected, uid })}
      id={uid}
      value={uid}
      name="batch"
      labelText=""
      handleChange={(isChecked) => {
        toggleCheckbox({
          isChecked,
          selected,
          setSelected,
          uid,
        })
      }}
    />
  )
}

// deleteSelected
// ============================================
/*
export const deleteSelected = ({ deleteApi, deletedUids, selected, setDeletedUids, setSelected }) => {
  // Add to deleted
  setDeletedUids([...deletedUids, ...selected])

  // Delete all items in selected
  selected.forEach(async (uid) => {
    await deleteApi(uid)
  })

  // Clear out selected
  setSelected([])
}
*/
export const deleteSelected = ({ batchConfirmationAction, deletedUids, selected, setDeletedUids, setSelected }) => {
  // Add to deleted
  setDeletedUids([...deletedUids, ...selected])

  // Call handler
  if (batchConfirmationAction) {
    batchConfirmationAction(selected)
  }

  // Clear out selected
  setSelected([])
}

// getIcon
// ============================================
export const getIcon = ({ field, sortOrder, sortBy, color = 'inherit' }) => {
  if (sortBy === field) {
    return sortOrder === 'asc' ? (
      // SORTED ASCENDING
      <KeyboardArrowUp color={color} />
    ) : (
      // SORTED DESCENDING
      <KeyboardArrowDown color={color} />
    )
  }
  // UNSORTED
  return <UnfoldMore color={color} />
}

// getSortState
// If sorted returns sortOrder (ie: 'asc' or 'desc')
// If not sorted returns undefined
// ============================================
export const getSortState = ({ field, sortOrder, sortBy }) => {
  if (sortBy === field) {
    return sortOrder
  }
  return undefined
}

// getUidsThatAreNotDeleted
// Returns all available uids that are not deleted
// Used to determine the state of the De/Select all button
// ============================================
export const getUidsThatAreNotDeleted = ({ computedDataUids = [], deletedUids = [] }) => {
  return computedDataUids.reduce((accumulator, computedDataUid) => {
    const isDeleted = deletedUids.find((deletedUid) => {
      return deletedUid === computedDataUid
    })
    return isDeleted === undefined ? [...accumulator, computedDataUid] : accumulator
  }, [])
}

// getUidsFromComputedData
// Returns array of uids from computedData
// Doesn't care about selected or deleted items
// ============================================
export const getUidsFromComputedData = ({ computedData = {} }) => {
  if (computedData && computedData.isFetching === false && computedData.success) {
    return computedData.response.entries.map((item) => {
      return item.uid
    })
  }
  return []
}

// isSelected
// ============================================
export const isSelected = ({ selected, uid }) => {
  const exists = selected.find((selectedUid) => {
    return selectedUid === uid
  })
  return exists ? true : false
}

// sortCol
// ============================================
/*
export const sortCol = ({ dialect_path, urlItemType, newSortBy, pageSize, pushWindowPath, siteTheme, sortOrder }) => {
  const url = `/${siteTheme}${dialect_path}/${urlItemType}/${pageSize}/1?sortBy=${newSortBy}&sortOrder=${
    sortOrder === 'asc' ? 'desc' : 'asc'
  }`
  NavigationHelpers.navigate(url, pushWindowPath, false)
}
*/
// sortCol
// ============================================
export const sortCol = ({ newSortBy, pageSize, pushWindowPath, sortOrder, sortHandler }) => {
  const page = 1
  const url = `${windowLocationPathnameWithoutPagination()}/${pageSize}/${page}`
  // Get search object, add in new sortBy & sortOrder
  const searchObj = Object.assign({}, getSearchObject(), {
    sortBy: newSortBy,
    sortOrder: sortOrder === 'asc' ? 'desc' : 'asc',
  })
  // Smash together & update url
  NavigationHelpers.navigate(`/${url}?${getSearchObjectAsUrlQuery(searchObj)}`, pushWindowPath, false)

  if (sortHandler) {
    sortHandler({
      page,
      pageSize,
      sortOrder: searchObj.sortOrder,
      sortBy: searchObj.sortBy,
    })
  }
}

// toggleCheckbox
// ============================================
export const toggleCheckbox = ({ isChecked, selected, setSelected, uid }) => {
  let _selected = [...selected]

  const exists = isSelected({ selected, uid })

  if (isChecked && !exists) {
    _selected.push(uid)
  }

  if (!isChecked && exists) {
    _selected = _selected.filter((selectedUid) => {
      return selectedUid !== uid
    })
  }

  setSelected(_selected)
}

// useDeleteItem
// ============================================
export const useDeleteItem = ({ deleteApi, deletedUids, deleteItemUid, selected, setDeletedUids, setSelected }) => {
  useEffect(() => {
    if (deleteItemUid) {
      // Add to deleted
      setDeletedUids([...deletedUids, deleteItemUid])

      // Remove from selected
      setSelected(
        // filter out deleteItemUid
        selected.filter((uid) => {
          return uid !== deleteItemUid
        })
      )

      // call server
      deleteApi(deleteItemUid)
    }
  }, [deleteItemUid])
}

// useGetData
// ============================================
export const useGetData = ({ computeData, dataPath, deletedUids, getData, routeParams, search }) => {
  useEffect(() => {
    getData()
  }, [routeParams.pageSize, routeParams.page, search.sortBy, search.sortOrder])

  // Filter out any items deleted after page load:
  const computedData = ProviderHelpers.getEntry(computeData, dataPath)
  if (computedData && computedData.isFetching === false && computedData.success) {
    const entries = computedData.response.entries
    const filtered = entries.reduce((accumulator, entry) => {
      const isDeleted = deletedUids.find((uid) => {
        return uid === entry.uid
      })
      if (isDeleted === undefined) {
        return [...accumulator, entry]
      }
      return accumulator
    }, [])
    const filterComputedData = Object.assign({}, computedData)
    filterComputedData.response.entries = filtered
    return filterComputedData
  }
  return computedData
}

// useGetCopy
// ============================================
export const useGetCopy = (webpackDynamicImport) => {
  const [copy, setCopy] = useState()
  useEffect(() => {
    const promised = webpackDynamicImport()
    promised.then((_copy) => {
      setCopy(_copy)
    })
  }, [])
  return copy
}

// usePaginationRequest
// ============================================
export const usePaginationRequest = ({ pushWindowPath, paginationRequest }) => {
  useEffect(() => {
    if (paginationRequest) {
      NavigationHelpers.navigate(paginationRequest, pushWindowPath, false)
    }
  }, [paginationRequest])
}
