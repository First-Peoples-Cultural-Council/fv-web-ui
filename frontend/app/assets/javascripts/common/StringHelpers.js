/*
Copyright 2016 First People's Cultural Council

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import IntlService from 'views/services/intl'

export const CLEAN_NXQL = 'NXQL'
export const CLEAN_FULLTEXT = 'fulltext'
export const CLEAN_ID = 'CLEAN_ID'

export default {
  clean: (str, mode = CLEAN_NXQL) => {
    let _str

    if (!str) return str

    if (mode === CLEAN_NXQL) {
      // Escape single quotes and URL decode
      _str = decodeURIComponent(str.replace(/'/g, "\\'"))

      // Escape '&' operator
      _str = _str.replace(/&/g, '%26')
    }

    if (mode === CLEAN_FULLTEXT) {
      // Escape single quotes and URL decode
      _str = str.replace(/'/g, "\\'")

      // Replace colon
      _str = _str.replace(/:/g, '\\:')

      // Escape double quotes
      _str = _str.replace(/"/g, '\\"')

      // Remove parentheses
      _str = _str.replace(/[()]/g, '')

      // Escape single quotes and URL decode
      _str = _str.replace(/!/g, '\\!')

      // Escape colon
      _str = decodeURIComponent(_str)
    }

    if (mode === CLEAN_ID) {
      // Swap : to -
      _str = str.replace(/:/g, '-')
      // Swap [] to empty string
      _str = _str.replace(/[[\]]/g, '')
    }

    return _str
  },
  extractErrorMessage: (jsonError) => {
    let errorMessage = jsonError.message

    if (jsonError.message !== null && jsonError.message.indexOf(': ') !== -1) {
      errorMessage = jsonError.message.split(': ')[1]
      errorMessage =
        IntlService.instance.translate({
          key: 'error',
          default: 'Error',
          append: ': ',
        }) + IntlService.instance.searchAndReplace(errorMessage)
    }

    return errorMessage
  },
  getReadableFileSize: (size) => {
    const e = (Math.log(size) / Math.log(1e3)) | 0 // TODO: Confirm use of `Bitwise OR`
    return +(size / Math.pow(1e3, e)).toFixed(2) + ' ' + ('kMGTPEZY'[e - 1] || '') + 'B'
  },
  toTitleCase: (string) => {
    return string[0].toUpperCase() + string.substring(1)
  },
  randomIntBetween: (min, max) => {
    return Math.floor(Math.random() * (Math.floor(max) - Math.ceil(min))) + Math.ceil(min)
  },
  formatUTCDateString: (dateString) => {
    let dayDesc
    const d = new Date(dateString)
    // let m = [
    //     intl.trans('jan', 'Jan', 'first'),
    //     intl.trans('feb', 'Feb', 'first'),
    //     intl.trans('mar', 'Mar', 'first'),
    //     intl.trans('apr', 'Apr', 'first'),
    //     intl.trans('may', 'May', 'first'),
    //     intl.trans('jun', 'Jun', 'first'),
    //     intl.trans('jul', 'Jul', 'first'),
    //     intl.trans('aug', 'Aug', 'first'),
    //     intl.trans('sept', 'Sep', 'first'),
    //     intl.trans('oct', 'Oct', 'first'),
    //     intl.trans('nov', 'Nov', 'first'),
    //     intl.trans('dec', 'Dec', 'first')
    // ];
    // @todo translate using the above?
    const m = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'June', 'July', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec']

    switch (d.getDate().toString()[d.getDate().toString().length - 1]) {
      case '1':
        dayDesc = 'st'
        break

      case '2':
        dayDesc = 'nd'
        break

      case '3':
        dayDesc = 'rd'
        break

      default:
        dayDesc = 'th'
    }

    return (
      m[d.getMonth()] +
      ' ' +
      d.getDate() +
      dayDesc +
      ', ' +
      d.getFullYear() +
      ' @ ' +
      ('0' + d.getUTCHours()).slice(-2) +
      ':' +
      ('0' + d.getUTCMinutes()).slice(-2)
    )
  },
  isUUID: (str) => {
    if (!str) {
      return false
    }

    return str.match(/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i)
  },
  queryStringToObject: (str, skipDecode = false) => {
    const _str = str.substring(str.indexOf('?') + 1).split('&')
    const params = {}
    let pair
    const d = skipDecode ? (s) => s : decodeURIComponent
    for (let i = _str.length - 1; i >= 0; i--) {
      pair = _str[i].split(/=(.+)/)
      params[d(pair[0])] = d(pair[1] || '')
    }

    return params
  },
  visibilityCategories: ({ visibility, dialectName = 'Community' }) => {
    switch (visibility) {
      case 'public':
        return 'Public'
      case 'team':
        return 'Language Team'
      case 'members':
        return `${dialectName} Members`
      default:
        return visibility
    }
  },
}
