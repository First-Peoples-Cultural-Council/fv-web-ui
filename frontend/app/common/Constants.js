export const STATE_LOADING = 0 // component is loading (eg: getting data) or busy (eg: submitting data)
export const STATE_DEFAULT = 1 // initial, loaded state, eg: displaying a form
export const STATE_ERROR = 2 // component is not happy, ie: form validation (not .js errors)
export const STATE_SUCCESS = 3 // component is happy, ie: form submitted
export const STATE_SUCCESS_DELETE = 5 // destructive action has succeeded
export const STATE_ERROR_BOUNDARY = 4 // component has triggered a .js error or is completely messed up

export const DEFAULT_PAGE = 1
export const DEFAULT_PAGE_SIZE = 100
export const DEFAULT_LANGUAGE = 'english'
export const DEFAULT_SORT_COL = 'dc:title'
export const DEFAULT_SORT_TYPE = 'asc'

export const WORKSPACES = 'Workspaces'
export const SECTIONS = 'sections'

export const UNKNOWN = 0
export const WORD = 1
export const PHRASE = 2
export const SONG = 3
export const STORY = 4
export const BOOK = 5

export const EVEN = 0
export const ODD = 1

export const URL_QUERY_PLACEHOLDER = 'first'

export const TABLE_FULL_WIDTH = 1
export const TABLEPAGINATION_DIV = 'TABLEPAGINATION_DIV'
export const WIDGET_WORKSPACE = 'WIDGET_WORKSPACE'

// --- Search ---
export const SEARCH_PART_OF_SPEECH_ANY = 'Any'
export const SEARCH_FILTERED_BY = 5
export const SEARCH_FILTERED_BY_CHARACTER = `${SEARCH_FILTERED_BY}.1`
export const SEARCH_FILTERED_BY_CATEGORY = `${SEARCH_FILTERED_BY}.2`
export const SEARCH_FILTERED_BY_PHRASE_BOOK = `${SEARCH_FILTERED_BY}.3`
// searchByMode CONSTANTS
export const SEARCH_BY_DEFAULT = 0 // TODO: REFACTOR TO DROP THIS
export const SEARCH_BY_CUSTOM = 3 // TODO: DROP THIS AFTER `app/components/__SearchDialect/SearchDialect.js` IS OBSOLETE

// USED WHEN SEARCHING ON DIFFERENT PAGES, EG: PHRASES VS WORDS
// GENERATES DIFFERENT UI TEXT
export const SEARCH_DATA_TYPE_PHRASE = 5
export const SEARCH_DATA_TYPE_WORD = 6

export const SEARCH_TYPE_APPROXIMATE_SEARCH = 'approx'
export const SEARCH_TYPE_EXACT_SEARCH = 'exact'
export const SEARCH_TYPE_CONTAINS_SEARCH = 'contains'
export const SEARCH_TYPE_STARTS_WITH_SEARCH = 'starts_with'
export const SEARCH_TYPE_ENDS_WITH_SEARCH = 'ends_with'
export const SEARCH_TYPE_WILDCARD_SEARCH = 'wildcard'
export const SEARCH_TYPE_DEFAULT_SEARCH = 'match'
export const SEARCH_TYPE_DEFAULT = SEARCH_TYPE_DEFAULT_SEARCH

export const SEARCHDIALECT_CHECKBOX = 'SEARCHDIALECT_CHECKBOX'
export const SEARCHDIALECT_SELECT = 'SEARCHDIALECT_SELECT'
