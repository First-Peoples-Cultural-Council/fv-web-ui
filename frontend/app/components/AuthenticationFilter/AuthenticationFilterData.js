import PropTypes from 'prop-types'

import useRoute from 'dataSources/useRoute'
import useLogin from 'dataSources/useLogin'

/**
 * @summary AuthenticationFilterData
 * @version 1.0.1
 * @component
 *
 * @param {object} props
 * @param {function} props.children
 *
 */
function AuthenticationFilterData({ children }) {
  const { routeParams } = useRoute()
  const { computeLogin } = useLogin()
  return children({
    routeParams,
    computeLogin,
  })
}
// PROPTYPES
const { func } = PropTypes
AuthenticationFilterData.propTypes = {
  children: func,
}

export default AuthenticationFilterData
