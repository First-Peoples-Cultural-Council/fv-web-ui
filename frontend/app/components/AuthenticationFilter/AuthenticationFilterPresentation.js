import React from 'react'
import PropTypes from 'prop-types'

import selectn from 'selectn'
import { SECTIONS } from 'common/Constants'
import Error403 from 'components/Error403'
import CircularProgress from '@material-ui/core/CircularProgress'

/**
 * @summary AuthenticationFilterPresentation
 * @version 1.0.1
 * @component
 *
 * @param {object} props
 *
 * @returns {node} jsx markup
 */
function AuthenticationFilterPresentation({
  // Calling components
  children,
  containerStyle,
  hideFromSections,
  is403,
  notAuthenticatedComponent,
  // Data layer
  routeParams,
  computeLogin,
}) {
  if (is403) {
    return <Error403 redirect={window.location.pathname} />
  }

  const { isFetching, success, isConnected } = computeLogin

  if (isFetching) {
    return <CircularProgress mode="indeterminate" />
  }

  // Logged in user.
  if (success && isConnected) {
    // Hide from sections for logged in user as well.
    const isSection = selectn('area', routeParams) === SECTIONS
    if (hideFromSections && isSection) {
      return null
    }

    return <div style={containerStyle}>{children}</div>
  }

  return notAuthenticatedComponent
}

// PROPTYPES
const { bool, node, object } = PropTypes
AuthenticationFilterPresentation.propTypes = {
  // Calling components
  children: node,
  containerStyle: object,
  hideFromSections: bool,
  is403: bool,
  notAuthenticatedComponent: node,
  // Data layer
  routeParams: object,
  computeLogin: object,
}
AuthenticationFilterPresentation.defaultProps = {
  computeLogin: {},
  notAuthenticatedComponent: null,
  is403: false,
  hideFromSections: false,
}
export default AuthenticationFilterPresentation
