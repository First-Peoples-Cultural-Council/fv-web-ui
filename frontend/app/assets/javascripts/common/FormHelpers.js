/*
Copyright 2016 First People's Cultural Council

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import selectn from 'selectn'
import React from 'react'
import StringHelpers from 'common/StringHelpers'
export default {
  // Get properties from form value
  getProperties: function(form) {
    const properties = {}
    const formValue = form.getValue()

    for (const key in formValue) {
      // Treat valued checkboxes differently. Always have value, so skip if unchecked.
      // getComponent does not work with input names that have '.' in them. Access directly.
      // valuedCheckbox = selectn('form.refs.input.refs[\'' + key + '\'].refs.valued_checkbox', form);

      let valuedCheckbox = null

      // Ensure Input for Form Ref is defined
      if (form.refs.input) {
        valuedCheckbox = form.refs.input.refs[key].refs.valued_checkbox
      }

      if (valuedCheckbox) {
        if (!valuedCheckbox.checked) {
          continue
        }
      }

      if (formValue.hasOwnProperty(key) && key) {
        if (formValue[key] && formValue[key] != '') {
          properties[key] = formValue[key]
        }
      }
    }

    return properties
  },
  prepareFilters: function(filters, options, optionsKey) {
    const preparedFilters = {}

    // Test each of the filters against item
    for (const filterKey in filters) {
      const filterOptions = selectn([optionsKey, 'fields', filterKey], options)

      // Add options to returned filter object

      // Filter not prepared
      if (!filters[filterKey].hasOwnProperty('appliedFilter')) {
        preparedFilters[filterKey] = {
          appliedFilter: filters[filterKey],
          filterOptions: filterOptions,
        }
      } else {
        preparedFilters[filterKey] = filters[filterKey]
      }
    }

    return preparedFilters
  },
}

/*
Get FormData from `form`

1) Pass in a reference to a `form`
2) Optionally provide regex of field names that should be converted to js objects (via `JSON.parse()`)

Example

getFormData({
  formReference: this.form,
  toParse: [
    /^fv:literal_translation/,
    /^fv:definitions/,
    /^fv:related_audio/,
    /^fv:related_pictures/,
    /^fv:cultural_note/,
    /^fvm:source/,
    /^fv-word:related_phrases/,
  ],
})
*/
export const getFormData = ({ formReference, toParse = [] }) => {
  const formDataFormatted = {}
  const formData = new FormData(formReference.current)

  for (const value of formData.entries()) {
    // parse any stringify-ed array/objects
    const name = value[0]
    const data = value[1]

    const shouldParse = toParse.some((regex) => {
      return regex.test(name)
    })

    if (shouldParse) {
      formDataFormatted[name] = JSON.parse(data)
      continue
    }

    // Handle file
    const isFile = data instanceof File
    if (isFile) {
      // Create empty array if needed:
      if (formDataFormatted[name] === undefined) {
        formDataFormatted[name] = []
      }
      // Get file array data:
      const fileData = formDataFormatted[name]
      // Update file array data only if a file exists:
      // console.log('data', data)
      if (data.name) {
        formDataFormatted[name] = [...fileData, data]
      }
      continue
    }

    formDataFormatted[name] = data
  }
  return formDataFormatted
}

/*

Validates a form

1) Provide FormData object of form and yup schema, optional: abortEarly boolean

Example

const validator = yup.object().shape({
      'dc:title': yup
        .string()
        .label('Name')
        .required()
    })
const formData = getFormData({
  formReference: this.refToForm,
})
const formValidation = await validateForm({formData, validator})

*/
export const validateForm = async({ formData, abortEarly = false, validator }) => {
  // Note: When `abortEarly === true` then `{ path, type } = invalid` is defined.
  // When `abortEarly === false` then `{ path, type } = invalid` is not defined! Data is found in `invalid.errors[]`.
  const validation = await validator.validate(formData, { abortEarly }).then(
    () => {
      return {
        valid: true,
        errors: [],
      }
    },
    (invalid) => {
      const { inner } = invalid
      const errors = inner.map((error) => {
        const { message, path, type } = error
        return {
          message,
          path,
          type,
        }
      })
      return {
        valid: false,
        errors,
      }
    }
  )
  return validation
}
/*
Determines if a field is valid. Returns object.

Example

const validator = yup.object().shape({
      'dc:title': yup
        .string()
        .label('Name')
        .required()
    })
const formData = getFormData({
  formReference: this.refToForm,
})
const formValidation = await validateForm({name: 'dc:title', formData, validator}})

Return

{
  valid: boolean,
  [yup error?]: [?]
}
*/
export const validateField = async({ name, formData, validator }) => {
  const results = await this._validateForm({ formData, validator })
  const { valid, errors } = results

  if (valid === false) {
    const fieldErrored = errors.filter((error) => {
      return error.path === name
    })
    if (fieldErrored.length !== 0) {
      const fieldData = fieldErrored[0]
      fieldData.valid = false
      return fieldData
    }
  }
  return {
    valid: true,
  }
}
/*
Returns a field's error message from yup validation

const validator = yup.object().shape({
  'dc:title': yup
    .string()
    .label('Name')
    .required(),
})
const formData = getFormData({
  formReference: this.refToForm,
})
const formValidation = await validateForm({ formData, validator })

if (formValidation.valid === false) {
  const errorForTitleField = getError({ errors: formValidation.errors, fieldName: 'dc:title' })
}
 */
export const getError = ({ errors = [], fieldName }) => {
  const error = errors.filter((errorItem) => {
    return errorItem.path === fieldName
  })
  if (error.length === 1) {
    return error[0]
  }
  return {}
}

export const handleSubmit = async({ validator, formData, valid, invalid }) => {
  const formValidation = await validateForm({ formData, validator })

  if (formValidation.valid) {
    valid()
  } else {
    invalid({
      errors: formValidation.errors,
    })
  }
}

export const getErrorFeedback = ({ errors = [] }) => {
  let errorFeedback = null
  if (errors.length !== 0) {
    if (errors.length > 0) {
      const li = errors.map((error, i) => {
        return (
          <li key={i}>
            <label className="Form__errorFeedbackItemLabel" htmlFor={StringHelpers.clean(error.path, 'CLEAN_ID')}>
              {error.message}
            </label>
          </li>
        )
      })

      const intro = `Could you review the following ${errors.length > 1 ? 'items' : 'item'}?`
      errorFeedback = (
        <div className="Form__errorFeedback">
          {intro}
          <ul className="Form__errorFeedbackItems">{li}</ul>
        </div>
      )
    }
  }
  return errorFeedback
}
