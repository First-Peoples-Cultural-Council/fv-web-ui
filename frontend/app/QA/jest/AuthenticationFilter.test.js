// Mock: react-redux
// --------------------------------------------------
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

// Let us begin!
// ==================================================
import React from 'react'
import renderer from 'react-test-renderer'
import AuthenticationFilterData from 'components/AuthenticationFilter/AuthenticationFilterData'
import AuthenticationFilterPresentation from 'components/AuthenticationFilter/AuthenticationFilterPresentation'
import { SECTIONS } from 'common/Constants'

describe('AuthenticationFilter', () => {
  // Data
  test('Data', () => {
    // Updated in AuthenticationFilterData.children
    let output = {}

    renderer.create(
      <AuthenticationFilterData>
        {(dataLayerOutput) => {
          output = dataLayerOutput
          return null
        }}
      </AuthenticationFilterData>
    )
    expect(output).toMatchInlineSnapshot(`
      Object {
        "computeLogin": undefined,
        "routeParams": undefined,
      }
    `)
  })

  // Presentation
  test('Presentation: render Error403', () => {
    const testProps = {
      is403: true,
      hideFromSections: false,
    }

    const component = renderer.create(<AuthenticationFilterPresentation {...testProps} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Presentation: render CircularProgress', () => {
    const testProps = {
      computeLogin: {
        isFetching: true,
      },
    }
    const component = renderer.create(<AuthenticationFilterPresentation {...testProps} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Presentation: render null', () => {
    const testProps = {
      computeLogin: {
        success: true,
        isConnected: true,
      },
      routeParams: {
        area: SECTIONS,
      },
      hideFromSections: true,
    }
    const component = renderer.create(<AuthenticationFilterPresentation {...testProps} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Presentation: render children', () => {
    const testProps = {
      computeLogin: {
        success: true,
        isConnected: true,
      },
      routeParams: {
        area: SECTIONS,
      },
      hideFromSections: false,
      children: 'renderedChildren',
    }
    const component = renderer.create(<AuthenticationFilterPresentation {...testProps} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Presentation: render notAuthenticatedComponent v1', () => {
    const testProps = {
      computeLogin: {
        success: true,
        isConnected: false,
      },
      notAuthenticatedComponent: <div>renderedNotAuthenticatedComponent</div>,
    }
    const component = renderer.create(<AuthenticationFilterPresentation {...testProps} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Presentation: render notAuthenticatedComponent v2', () => {
    const testProps = {
      computeLogin: {
        success: false,
      },
      notAuthenticatedComponent: <div>renderedNotAuthenticatedComponent</div>,
    }
    const component = renderer.create(<AuthenticationFilterPresentation {...testProps} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
