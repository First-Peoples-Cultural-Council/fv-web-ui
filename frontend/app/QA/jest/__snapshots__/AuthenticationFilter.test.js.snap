// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationFilter Presentation: render CircularProgress 1`] = `
<div
  className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
  mode="indeterminate"
  role="progressbar"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
>
  <svg
    className="MuiCircularProgress-svg"
    viewBox="22 22 44 44"
  >
    <circle
      className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
      cx={44}
      cy={44}
      fill="none"
      r={20.2}
      strokeWidth={3.6}
      style={Object {}}
    />
  </svg>
</div>
`;

exports[`AuthenticationFilter Presentation: render Error403 1`] = `
<div
  className="Error403"
>
  <h1
    className="Error403__heading"
  />
  <div>
    <p />
    <p>
      <a
        className="RaisedButton RaisedButton--primary"
        href="logout?requestedUrl=../"
        style={
          Object {
            "alignItems": "center",
            "color": "#fff",
            "display": "inline-flex",
            "justifyContent": "center",
            "textDecoration": "none",
          }
        }
      />
    </p>
  </div>
</div>
`;

exports[`AuthenticationFilter Presentation: render children 1`] = `
<div>
  renderedChildren
</div>
`;

exports[`AuthenticationFilter Presentation: render notAuthenticatedComponent v1 1`] = `
<div>
  renderedNotAuthenticatedComponent
</div>
`;

exports[`AuthenticationFilter Presentation: render notAuthenticatedComponent v2 1`] = `
<div>
  renderedNotAuthenticatedComponent
</div>
`;

exports[`AuthenticationFilter Presentation: render null 1`] = `null`;
